cpssEnabler/linuxNoKernelModule/drivers/dmaDriver.c

```diff
@@ -28,320 +28,929 @@ disclaimer.
+						
+						mb();
+						
+						reset_done++;
+						
+						} 
+						else pr_info("%s: base %d BAR4 not mappable!\n", __func__, i);
+					
+        }
+	else pr_info("%s: unable to iomap %d. err %d!", __func__, i, err);
+ 	}
+
+	 udelay(20);
+	 
+ 	for (i=0; i<2; i++) {
+		pr_info("%s: Stopped DMA on PCI device %x:%x:%x.%x\n", __func__, domain, bus + i, (unsigned)((off >> 3) & 0x1f), (unsigned)(off & 0x07));
+		/*printk("%s: device %d: wrote %x to DFX PCIe skip configuration register %x\n", 
+				__func__, i, skip_pcie_reg[i], DFX_PCIE_SKIP_INIT_MATRIX_REG);*/
+		/*printk("%s: device %d: Read %x from PCIe configuration register %x\n", 
+				__func__, i, pcie_conf_hdr[i], CNM_PCIE_CONF_HDR_REG);*/
+		if (reset_done)
+		   printk("%s: device %d: Wrote %x to DFX rst reg %x reset_cnt %d\n", 
+				   __func__, i, dfx_rst_reg[i], DFX_RST_CTRL_REG, reset_done);
+
+ 		}
+  	/*	if (m->base[i]) {
+               pr_info("%s: Stopped DMA on PCI device %x:%x:%x.%x\n", __func__, domain, bus + i, (unsigned)((off >> 3) & 0x1f), (unsigned)(off & 0x07)); 
+
+				did = readl(m->base[i] + MG_DEVID_REG);
+				vid = readl(m->base[i] + MG_VENDID_REG);
+				pr_info("%s: did:vid %x:%x\n", __func__, did, vid);
+	
+				 val = readl(m->base[i] + MG_UEA_REG);
+				pr_info("%s: UEA value is: %x\n", __func__, val);  Unit Error Address (UEA) Register 
+
+				 Finally, take SDMA out of reset: 
+				val = readl(m->base[i] + MG_EXT_GLOBAL_CNTRL_REG);
+				val &= (~SDMA_SW_RST_BIT);
+				writel(val, m->base[i] + MG_EXT_GLOBAL_CNTRL_REG);
+	
+  			}
+  	}*/
+
+	return err;
+
 }
 
 static int mvDmaDrv_release(struct inode *inode, struct file *file)
 {
-    struct dma_mapping *m = (struct dma_mapping *)file->private_data;
-    printk("mvDmaDrv_release(file=%p) data=%p\n", file, m);
-    if (m != shared_dmaBlock) {
-        free_dma_block(m);
-        kfree(m);
-    }
-    return 0;
+	struct dma_mapping *m = (struct dma_mapping *)file->private_data;
+	int ret;
+
+	printk("%s: %s(file=%p) data=%p\n", MV_DRV_NAME, __func__, file, m);
+	ret = mvDmaDrv_stopAndReset_AC3X_Aldrin_PP(m);
+
+	printk("%s: %s: return value of stop and reset PP SDMA RX is: %d\n", 
+			MV_DRV_NAME, __func__, ret);
+	printk("%s: delaying DMA memory free by %d second...\n", __func__, SECS_DELAY_WQ);
+	schedule_delayed_work(&m->free_mem_delayed, msecs_to_jiffies(1000*SECS_DELAY_WQ));
+
+	return ret;
+}
+
+static int mvdmadrv_pdriver_probe(struct platform_device *pdev)
+{
+	int err;
+
+	err = of_reserved_mem_device_init(&pdev->dev);
+	if (err) {
+		dev_err(&pdev->dev, "Could not get reserved memory\n");
+		return -ENOMEM;
+	}
+
+	platdrv_dev = &pdev->dev;
+
+	printk("%s: Using platform device %s\n", MV_DRV_NAME, pdev->name);
+
+	return 0;
+};
+
+static int mvdmadrv_pdriver_remove(struct platform_device *pdev)
+{
+	BUG_ON(!platdrv_registered);
+
+	of_reserved_mem_device_release(&pdev->dev);
+
+	return 0;
 }
 
+static const struct of_device_id mvdmadrv_of_match_ids[] = {
+	 { .compatible = "marvell,mv_dma", },
```

Comment:
The driver initializes the character device interface regardless. It also registers itself at the end as a platform driver, but this has little effect since all of the initialization was already performed (e.g. its initialization is not dependent on dtb matching of ids)